ArcGISRuntimeSDKQt_CppSamples/Analysis/PerformValveIsolationTrace/PerformValveIsolationTrace.cpp

```diff
@@ -0,0 +1,273 @@
+}
+
+// Set the view (created in QML)
+void PerformValveIsolationTrace::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  // apply renderers
+  SimpleMarkerSymbol* startingPointSymbol = new SimpleMarkerSymbol(SimpleMarkerSymbolStyle::Cross, Qt::green, 25, this);
+  m_startingLocationOverlay->setRenderer(new SimpleRenderer(startingPointSymbol, this));
+  m_mapView->graphicsOverlays()->append(m_startingLocationOverlay);
+
+  emit mapViewChanged();
+}
+
+QStringList PerformValveIsolationTrace::categoriesList() const
+{
+  if (!m_utilityNetwork)
+    return { };
+
+  if (m_utilityNetwork->loadStatus() != LoadStatus::Loaded)
+    return { };
+
+  QList<UtilityCategory*> categories = m_utilityNetwork->definition()->categories();
+  QStringList strList;
+  for (UtilityCategory* category : categories)
+  {
+    strList << category->name();
+  }
+  return strList;
+}
+
+void PerformValveIsolationTrace::performTrace()
+{
+  if (m_selectedIndex < 0)
+    return;
+
+  // disable UI while trace is run
+  m_uiEnabled = false;
+  emit uiEnabledChanged();
+
+  for (Layer* layer : *m_map->operationalLayers())
+  {
+     // clear previous selection from the feature layers
+    if (FeatureLayer* featureLayer = dynamic_cast<FeatureLayer*>(layer))
+    {
+      featureLayer->clearSelection();
+    }
+  }
+
+  QList<UtilityCategory*> categories = m_utilityNetwork->definition()->categories();
+
+  // get the selected utility category
+  if (categories[m_selectedIndex] != nullptr)
+  {
+    UtilityCategory* selectedCategory = categories[m_selectedIndex];
+    UtilityCategoryComparison* categoryComparison = new UtilityCategoryComparison(selectedCategory, UtilityCategoryComparisonOperator::Exists, this);
+
+    // set the category comparison to the barriers of the configuration's trace filter
+    m_traceConfiguration->filter()->setBarriers(categoryComparison);
+
+    // set whether to include isolated features
+    m_traceConfiguration->setIncludeIsolatedFeatures(m_isolateFeatures);
+
+    // build parameters for the isolation trace
+    UtilityTraceParameters* traceParameters = new UtilityTraceParameters(UtilityTraceType::Isolation, QList<UtilityElement*> {m_startingLocation}, this);
+    traceParameters->setTraceConfiguration(m_traceConfiguration);
+
+    m_utilityNetwork->trace(traceParameters);
+  }
+}
+
+void PerformValveIsolationTrace::connectSignals()
+{
+  connect(m_utilityNetwork, &UtilityNetwork::doneLoading, this, [this](const Error& error)
+  {
+    if (!error.isEmpty())
+    {
+      qDebug() << error.message() << error.additionalMessage();
+      return;
+    }
+
+    if (m_utilityNetwork->loadStatus() != LoadStatus::Loaded)
+      return;
+
+    // get a trace configuration from a tier
+    UtilityNetworkDefinition* networkDefinition = m_utilityNetwork->definition();
+    UtilityDomainNetwork* domainNetwork = networkDefinition->domainNetwork("Pipeline");
```

Comment:
do we need to test if any of these types obtained with a hard coded string are `nullptr`?